### PR TITLE
The Command Engine

### PR DESCRIPTION
**Pull request by Davarice**
The Big Overhaul to the commands system, six months in the making. Package-based command routing for segmentation of code and standardization of authentication.

- Commands are now parsed by way of shlex splitting, allowing space separated arguments except where quoted.

- Commands now take a rudimentary implementation of C-style options and flags.

- All commands now have reworked and dynamically generated help text which automatically uses the correct command prefix.

### CODE REVIEW

File: petal/commands/listener.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+
+class CommandsListener(core.Commands):
+    auth_fail = "This command requires the `{role}` role."
+    role = "RoleListener"
+
+    def authenticate(self, src):
+        return self.check_user_has_role(src.author, self.config.get(self.role))
+
+    async def cmd_lpromote(self, src, user=None, **_):
```
Comment (hdmifish) [inline petal/commands/listener.py]: We could probably get rid of this command, yeah?

File: petal/commands/listener.py [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+            return (
+                "A vote to promote {0}#{1} has been started, it will end in 48 hours.".format(
+                    user.name, user.discriminator
+                )
+                + "\nYou man cancel this vote by running "
+                + self.config.prefix
+                + "lcancel (not to be confused with smash bros)"
+            )
+
+    async def cmd_ldemote(self, src, user=None, **_):
```
Comment (hdmifish) [inline petal/commands/listener.py]: see above comment

File: petal/commands/listener.py [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+            return (
+                "A vote to demote {0}#{1} has been started, it will end in 48 hours.".format(
+                    user.name, user.discriminator
+                )
+                + "\nYou may cancel this vote by running "
+                + self.config.prefix
+                + "lcancel (not to be confused with smash bros)"
+            )
+
+    async def cmd_lcancel(self, src, **_):
```
Comment (hdmifish) [inline petal/commands/listener.py]: see above comment

File: petal/commands/listener.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+                temp[entry] = cb[entry]
+
+        for e in temp:
+            print("Deleted: " + str(self.config.doc["choppingBlock"][e]))
+            del self.config.doc["choppingBlock"][e]
+
+        self.config.save()
+        return "Deleted all lvotes if you had started any"
+
+    async def cmd_lvalidate(self, src, user=None, **_):
```
Comment (hdmifish) [inline petal/commands/listener.py]: see above comment

File: petal/commands/listener.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+                            "\nLet them know to type !Laccept in a channel"
+                        )
+                    else:
+                        return (
+                            user.name
+                            + " has been made a Helping Hands!\nThey must accept"
+                            " the invite by following the instruction I just sent them"
+                        )
+
+    async def cmd_laccept(self, src, **_):
```
Comment (hdmifish) [inline petal/commands/listener.py]: see above comment

File: petal/commands/listener.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+                        "Just letting y'all know that "
+                        + member.name
+                        + " has accepted their role",
+                    )
+
+                return "Welcome!"
+            else:
+                return "You don't have a pending invite to join the Helping Hands at this time"
+
+    async def cmd_lreject(self, src, **_):
```
Comment (hdmifish) [inline petal/commands/listener.py]: see above comment

File: petal/commands/listener.py [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+                        + src.author.name
+                        + " has rejected their role",
+                    )
+                del self.config.doc["choppingBlock"][src.author.id]
+                self.config.save()
+                return "You have rejected to join the helping hands. If this was on accident, let a listener know"
+            else:
+                return "You don't have a pending invite to join the Helping Handss at this time"
+
+    async def cmd_lshow(self, src, **_):
```
Comment (hdmifish) [inline petal/commands/listener.py]: see above comment

**Comment by Davarice** [comment]
I think that decision ultimately falls to LM, but I have no problem removing them.

**Comment by hdmifish** [comment]
Waiting on them to respond. May be a bit.

**Comment by hdmifish** [comment]
They've confirmed they want to keep it

**Pull request merged**